File: simgrid-template/MpiEnv/simgrid/Simgrid-git/examples/lua/splaySim/master.lua
-- Copyright (c) 2011, 2014. The SimGrid Team.
-- All rights reserved.

-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the license (GNU LGPL) which comes with this package.

--Master Function
function Master(...) 

simgrid.info("Hello from lua, I'm the master")
for i,v in ipairs(arg) do
    simgrid.info("Got "..v)
end

nb_task = arg[1];
comp_size = arg[2];
comm_size = arg[3];
slave_count = arg[4]

if (#arg ~= 4) then
    error("Argc should be 4");
end
simgrid.info("Argc="..(#arg).." (should be 4)")

-- Dispatch the tasks

for i=1,nb_task do
  tk = simgrid.task.new("Task "..i,comp_size,comm_size);
  alias = "slave "..(i%slave_count);
  simgrid.info("Master sending  '" .. simgrid.task.get_name(tk) .."' To '" .. alias .."'");
  simgrid.task.send(tk,alias); -- C user data set to NULL
  simgrid.info("Master done sending '".. simgrid.task.get_name(tk) .."' To '" .. alias .."'");
end

-- Sending Finalize Message To Others

simgrid.info("Master: All tasks have been dispatched. Let's tell everybody the computation is over.");
for i=0,slave_count-1 do
  alias = "slave "..i;
  simgrid.info("Master: sending finalize to "..alias);
  finalize = simgrid.task.new("finalize",comp_size,comm_size);
  simgrid.task.send(finalize,alias)
end
  simgrid.info("Master: Everything's done.");
end

-- end_of_master

File: simgrid-template/MpiEnv/simgrid/Simgrid-git/examples/scala/master_slave_bypass/Master.scala
/*
 * Master of a basic master/slave example in Java
 *
 * Copyright (c) 2006-2013. The SimGrid Team.
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package.
 */

package master_slave_bypass

import Stream._
import org.simgrid.msg.HostNotFoundException
import org.simgrid.msg.Msg
import org.simgrid.msg.MsgException
import org.simgrid.msg.Process
import org.simgrid.msg.Task

class Master(hostname:String, name:String) extends Process(hostname,name) {

  def main(args:Array[String]) {
    Msg.info("Master Hello!")

    //Create a slave on host "alice"
    try {
      Msg.info("Create process on host 'Tremblay'")
      new Slave("Tremblay","process2").start()
    } catch {
      case e:MsgException => println("Process2!")
    }

    //Wait for slave "alice"
    continually({Task.receive("Tremblay")})
      .takeWhile(!_.isInstanceOf[FinalizeTask])
      .force // to force the list to be compute

    Msg.info("Received Finalize. I'm done. See you!")
  }
}
